File: 백준/Silver/9012. 괄호/괄호.py
import sys

T = int(sys.stdin.readline())


for _ in range(T):
    D = sys.stdin.readline().strip()
    stack = []
    for p in D:
        if p == '(':
            stack.append(p)
        else:
            if len(stack):
                stack.pop()
            else:
                print('NO')
                break
    else:
        if stack:
            print('NO')
        else:
            print('YES')

File: 백준/Silver/9012. 괄호/괄호.cc
#include <iostream>
#include <stack>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);


    int n;
    cin>>n;
    for(int i=0;i<n;i++){
        stack<char> st;
        string state="YES";
        string input;
        cin>>input;


        for(int j=0;j<input.length();j++)
        {
            if(input[j]=='(')
                st.push(input[0]);
            else if(input[j]==')'){
                if(st.empty()){
                    state="NO";
                    break;}
                else
                    st.pop();
            }
        }

    if(st.empty())
        cout<<state<<'\n';
    else
        cout<<"NO"<<'\n';}
    return 0;
}

